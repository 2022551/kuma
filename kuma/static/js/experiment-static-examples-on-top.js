(function() {
    'use strict';

    var lou = new Mozilla.TrafficCop({
        id: 'experiment-static-examples-on-top',
        cookieExpires: 24 * 365, // 1 year
        variations: {
            'v=control': 50, // original, archived ad Experiment:StaticExamplesOnTop
            'v=example': 50  // Static example at the top of the page, at "normal" location
        }
    });
    lou.init();

})(window.Mozilla);
