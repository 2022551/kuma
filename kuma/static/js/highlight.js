(function($, win) {
    'use strict';

    // find all targets and wrap appropriately
    function highlight(targets) {
        var $targets = $(targets);
        $targets.each(function() {
            $(this).addClass('highlight-spanned').wrapInner('<span class="highlight-span"></span>');
        });
    }

    var $articleSubHeads;
    // call on aritcle body with targets
    if(!win.waffle || !win.waffle.flag_is_active('line_length')) {
        $articleSubHeads = $('#wikiArticle h2');
    } else {
        $articleSubHeads = $('#wikiArticle h3, #wikiArticle h5');
    }
    highlight($articleSubHeads);

})(jQuery, window);
