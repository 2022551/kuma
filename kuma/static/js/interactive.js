(function() {
    'use strict';

    var iframe = document.querySelector('iframe.interactive');
    var mediaQuery = window.matchMedia('(min-width: 63.9385em)');
    var targetOrigin =
        window.mdn.interactiveEditor.editorUrl ||
        'https://interactive-examples.mdn.mozilla.net';

    /**
     * Generate and returns a random string thanks to:
     * https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript
     */
    function randomString() {
        var text = '';
        var possible =
            'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

        for (var i = 0; i < 5; i++) {
            text += possible.charAt(
                Math.floor(Math.random() * possible.length)
            );
        }

        return text;
    }

    /**
     * Handles postMessages from the interactive editor, passing
     * the data in the relevant format to trackEvent depending on the label.
     * Also ensures that the origin of the message is what we expect.
     * @param {Object} event - The event Object received from the postMessage
     */
    function interactiveExamplesMsgHandler(event) {
        var allowedOrigin =
            window.mdn.interactiveEditor.editorUrl ||
            'https://interactive-examples.mdn.mozilla.net';

        if (event.origin !== allowedOrigin) {
            return false;
        }

        if (event.data.label === 'Performance Events') {
            mdn.analytics.trackEvent({
                category: event.data.category,
                action: event.data.action,
                label: new Date().getTime() + '-' + randomString(),
                value: event.data.value - performance.timing.fetchStart
            });
        } else {
            mdn.analytics.trackEvent(event.data);
        }
    }

    /**
     * Returns the interactive examples iframe performance entry
     * @param {Array} perfEntries - The array of performance entries
     */
    function getInteractiveExamplesPerfEntry(perfEntries) {
        for (var i = 0, l = perfEntries.length; i < l; i++) {
            var currentEntry = perfEntries[i];
            if (
                currentEntry.initiatorType &&
                currentEntry.initiatorType === 'iframe' &&
                currentEntry.name.indexOf(
                    'interactive-examples.mdn.mozilla.net'
                ) > -1
            ) {
                return perfEntries[i];
            }
        }
    }

    /* Ensure there is an iframe present and that performance is
       supported before gathering performance metric */
    if (iframe && performance !== undefined) {
        document.addEventListener('readystatechange', function(event) {
            if (event.target.readyState === 'complete') {
                var interactiveExamplesPerfEntry = {};
                var perfEntries = performance.getEntriesByType('resource');
                var mainFetchStart = 0;
                var iframeFetchStart = 0;
                var iframeFetchStartSinceUnixEpoch = 0;
                var timeToIframeFetchStart = 0;

                if (perfEntries === undefined || perfEntries.length <= 0) {
                    console.info('No performance entries was returned');
                    return;
                }

                interactiveExamplesPerfEntry = getInteractiveExamplesPerfEntry(
                    perfEntries
                );

                // ensure that the iframe was found in the array of performance entries
                if (interactiveExamplesPerfEntry !== undefined) {
                    mainFetchStart = performance.timing.fetchStart;
                    iframeFetchStart = Math.round(
                        interactiveExamplesPerfEntry.fetchStart
                    );
                    iframeFetchStartSinceUnixEpoch =
                        mainFetchStart + iframeFetchStart;

                    timeToIframeFetchStart =
                        new Date(iframeFetchStartSinceUnixEpoch) -
                        new Date(mainFetchStart);

                    mdn.analytics.trackEvent({
                        category: 'Interactive Examples',
                        action: 'Time to iframe fetch start',
                        label: new Date().getTime() + '-' + randomString(),
                        value: timeToIframeFetchStart
                    });
                }
            }
        });
    }

    /* If there is no `iframe`, or if this is a JS example,
    simply return */
    if (!iframe || iframe.classList.contains('interactive-js')) {
        return;
    }

    /**
     * A simple wrapper function for the `postMessage`s sent
     * to the interactive editor iframe
     * @param {Object} message - The message object sent to the interactive editor
     */
    function postToEditor(message) {
        iframe.contentWindow.postMessage(message, targetOrigin);
    }

    /* As the user sizes the browser or tilts their device,
    listen for mediaQuery events and communicate the new
    viewport state to the interactive editor */
    mediaQuery.addListener(function(event) {
        if (event.matches) {
            postToEditor({ smallViewport: false });
        } else {
            postToEditor({ smallViewport: true });
        }
    });

    window.onload = function() {
        // if the mediaQuery does not match on load
        if (!mediaQuery.matches) {
            // add the class `small-desktop-and-below`
            postToEditor({ smallViewport: true });
        }
    };

    // add event listener for postMessages from the interactive editor
    window.addEventListener('message', interactiveExamplesMsgHandler, false);
})();
