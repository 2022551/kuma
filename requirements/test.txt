# Requirements to run functional tests

-c constraints.txt
-r default_and_test.txt

# Implement bash brace expansion
braceexpand==0.1.1 \
    --hash=sha256:f967ca39bdb98e16299a69c45a944c5d4345393615ed6470bb1e62ca3506bf41

# Page Object Model for testing with Selenium
PyPOM==1.0 \
    --hash=sha256:a8728981970404c3c9a14efeb23da92191749a4a4ad489130fd855e4cf8f154e \
    --hash=sha256:5c98e3d8ef2c7f96c3365a3b8295560446f2447eecdcfe393a30abf2fcc16d96


# Test plugin: Re-run flaky tests
pytest-rerunfailures==2.1.0 \
    --hash=sha256:e867cec5eabb20ed38e04b6b9c277e34b0603143ee315beab56296d8329fe3f2 \
    --hash=sha256:82e6cd823c50ff2d1b2b183642302d42c1650bcf387b17d46f5711e08fa0995f

# Test plugin: Run tests with selenium
# Code: https://github.com/pytest-dev/pytest-selenium
# Changes: http://pytest-selenium.readthedocs.io/en/latest/news.html
# Docs: http://pytest-selenium.readthedocs.io/en/latest/index.html
pytest-selenium==1.11.4 \
    --hash=sha256:b66651fe7cbeee02b511f7b59f250ca77fcdb6024f193ca10da27d1d91240688 \
    --hash=sha256:9a0c48c434b538387ed6fa9d0c2f0b2e32f4fb71a4c41754df49be0aa4c64ae1

# Test plugin: Run tests in parallel
pytest-xdist==1.16.0 \
    --hash=sha256:42e5a1e5da9d7cff3e74b07f8692598382f95624f234ff7e00a3b1237e0feba2

