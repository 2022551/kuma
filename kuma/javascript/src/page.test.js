//@flow
import React from 'react';
import { create } from 'react-test-renderer';
import DocumentProvider from './document-provider.jsx';
import { fakeDocumentData } from './document-provider.test.js';
import Page from './page.jsx';

test('Page snapshot', () => {
    const page = create(
        <DocumentProvider initialDocumentData={fakeDocumentData}>
            <Page />
        </DocumentProvider>
    );
    expect(page.toJSON()).toMatchSnapshot();
});
