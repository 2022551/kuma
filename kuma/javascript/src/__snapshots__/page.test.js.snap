// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page snapshot 1`] = `
.emotion-38 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-top: 5px solid #83d0f2;
}

.emotion-0 {
  display: block;
  height: 48px;
  width: 219px;
  overflow: hidden;
}

.emotion-1 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 4px;
  -ms-flex-preferred-size: 4px;
  flex-basis: 4px;
}

.emotion-7 {
  position: relative;
  pointer-events: auto;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 18px;
  font-weight: bold;
  line-height: 48px;
  white-space: nowrap;
  padding: 0 10px;
  margin: 0;
  border: none;
}

.emotion-5:hover {
  background-color: #eee;
}

.emotion-5:focus {
  outline-offset: -3px;
}

.emotion-3 {
  font-size: 12px;
  padding-left: 2px;
}

.emotion-27 {
  -webkit-flex: 1 1 0px;
  -ms-flex: 1 1 0px;
  flex: 1 1 0px;
}

.emotion-31 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-bottom: solid black 1px;
  max-width: 350px;
  min-width: 120px;
  -webkit-flex: 2 1 120px;
  -ms-flex: 2 1 120px;
  flex: 2 1 120px;
}

.emotion-29 {
  vertical-align: -2px;
  -webkit-flex: 0 0 20px;
  -ms-flex: 0 0 20px;
  flex: 0 0 20px;
}

.emotion-30 {
  border-width: 0 !important;
  font-size: 18px;
  font-weight: bold !important;
  padding-bottom: 0 !important;
  -webkit-flex: 0 1 300px;
  -ms-flex: 0 1 300px;
  flex: 0 1 300px;
}

.emotion-32 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 15px;
  -ms-flex-preferred-size: 15px;
  flex-basis: 15px;
}

.emotion-35 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 18px;
  font-weight: bold;
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
  line-height: 48px;
  padding: 0 8px;
}

.emotion-35:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #eee;
}

.emotion-35:focus {
  outline-offset: -3px;
}

.emotion-34 {
  margin-left: 5px;
}

.emotion-41 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f5f9fa;
  padding: 16px 0 16px 32px;
}

.emotion-40 {
  font-family: x-locale-heading-primary,zillaslab,"Palatino","Palatino Linotype",x-locale-heading-secondary,serif;
  font-size: 2.8rem;
  margin: 0;
}

.emotion-43 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #fff;
  background-color: #222;
  padding: 15px 0 15px 32px;
}

.emotion-43 li {
  display: inline-block;
}

.emotion-43 a {
  color: #83d0f2;
  padding-right: 30px;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
}

.emotion-50 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  padding: 30px 24px;
}

.emotion-46 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-47 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 3%;
  -ms-flex-preferred-size: 3%;
  flex-basis: 3%;
}

.emotion-49 {
  -webkit-flex: 3;
  -ms-flex: 3;
  flex: 3;
}

.emotion-49 p {
  max-width: 42rem;
}

<div>
  <div
    className="emotion-38 emotion-39"
  >
    <a
      className="emotion-0"
      href="/en-US/"
    >
      <img
        alt="MDN Web Docs"
        src="/static/img/web-docs-sprite.svg"
      />
    </a>
    <div
      className="emotion-1 emotion-2"
      width={4}
    />
    <div
      className="emotion-7 emotion-8"
    >
      <button
        className="emotion-5 emotion-6"
        onClick={[Function]}
      >
        Technologies
        <span
          className="emotion-3 emotion-4"
        >
          ▼
        </span>
      </button>
    </div>
    <div
      className="emotion-7 emotion-8"
    >
      <button
        className="emotion-5 emotion-6"
        onClick={[Function]}
      >
        References & Guides
        <span
          className="emotion-3 emotion-4"
        >
          ▼
        </span>
      </button>
    </div>
    <div
      className="emotion-7 emotion-8"
    >
      <button
        className="emotion-5 emotion-6"
        onClick={[Function]}
      >
        Feedback
        <span
          className="emotion-3 emotion-4"
        >
          ▼
        </span>
      </button>
    </div>
    <div
      className="emotion-7 emotion-8"
    >
      <button
        className="emotion-5 emotion-6"
        onClick={[Function]}
      >
        <svg mock />
        <span
          className="emotion-3 emotion-4"
        >
          ▼
        </span>
      </button>
    </div>
    <div
      className="emotion-27 emotion-28"
    />
    <form
      action="/en-US/search"
      className="emotion-31"
      id="nav-main-search"
      method="get"
      role="search"
    >
      <svg mock
        className="emotion-29"
      />
      <input
        className="emotion-30"
        id="main-q"
        name="q"
        placeholder="Search MDN"
        type="search"
      />
    </form>
    <div
      className="emotion-32 emotion-2"
      width={15}
    />
    <a
      className="emotion-35"
      data-service="GitHub"
      href="/users/github/login/?next=/"
      rel="nofollow"
    >
      Sign in
       
      <svg mock
        className="emotion-34"
      />
    </a>
    <div
      className="emotion-32 emotion-2"
      width={15}
    />
  </div>
  <div
    className="emotion-41 emotion-39"
  >
    <h1
      className="emotion-40"
    >
      [fake document title]
    </h1>
  </div>
  <div
    className="emotion-43 emotion-39"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<ol>[fake TOC HTML]</ol>",
      }
    }
  />
  <div
    className="emotion-50 emotion-39"
  >
    <div
      className="emotion-46"
    >
      <nav
        className="crumbs"
        role="navigation"
      >
        <ol>
          <li
            className="crumb"
          >
            <a
              href="[fake grandparent url]"
            >
              <span>
                [fake grandparent title]
              </span>
            </a>
          </li>
          <li
            className="crumb"
          >
            <a
              href="[fake parent url]"
            >
              <span>
                [fake parent title]
              </span>
            </a>
          </li>
          <li
            className="crumb"
          >
            <span>
              [fake document title]
            </span>
          </li>
        </ol>
      </nav>
      <div
        className="quick-links emotion-45"
      >
        <div
          className="quick-links-head"
        >
          Related Topics
        </div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "[fake quicklinks HTML]",
            }
          }
        />
      </div>
    </div>
    <div
      className="emotion-47 emotion-2"
      width="3%"
    />
    <article
      className="text-content emotion-49"
      dangerouslySetInnerHTML={
        Object {
          "__html": "[fake body HTML]",
        }
      }
    />
  </div>
</div>
`;
